crazy_rl/multi_agent/jax/base_parallel_env.py

```diff
@@ -0,0 +1,151 @@
+        _action_space: Returns the Space object corresponding to valid actions
+        _observation_space: Returns the Space object corresponding to valid observations
+        _compute_obs: Computes the current observation of the environment from a given state.
+        _transition_state: Transitions the state based on the mechanics of the environment, for example makes the
+                           target move and sanitize the actions.
+        _sanitize_action: Makes the actions passed to step fit the environment, e.g. avoid making brutal moves.
+        _compute_reward: Computes the current reward value(s) from a given state.
+        _compute_terminated: Computes if the game must be stopped because the agents crashed from a given state.
+        _compute_truncation: Computes if the game must be stopped because it is too long from a given state.
+        _initialize_state: Initialize the State of the environment.
+        reset: Resets the environment in initial state.
+        auto_reset: Returns the State reinitialized if needed, else the actual State.
+        state_to_dict: Translates the State into a dict.
+        step_vmap: Used to vmap step, takes the values of the state and calls step with a new State object containing
+                   the state values.
+        state: Returns a global observation (concatenation of all the agent locations and the target locations).
+
+    There are also the following functions:
+        observation_space: Returns the observation space for one agent.
+        action_space: Returns the action space for one agent.
+    """
+
+    def _observation_space(self, agent: int) -> spaces.Space:
+        """Returns the observation space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _action_space(self, agent: int) -> spaces.Space:
+        """Returns the action space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_obs(self, state: State) -> State:
+        """Computes the current observation of the environment from a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _transition_state(self, state: State, actions: jnp.ndarray, key: jnp.ndarray) -> State:
+        """Transitions the state based on the mechanics of the environment, for example makes the target move and sanitize the actions. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _sanitize_action(self, state: State, actions: jnp.ndarray) -> State:
+        """Makes the actions passed to step fit the environment, e.g. avoid making brutal moves. Must be implemented in a subclass.
+
+        Args:
+            state : the state of the environment (contains agent_location used in this function).
+            actions : 2D array containing the x, y, z action for each drone.
+        """
+        raise NotImplementedError
```

Comment:
I think the implementation can be moved to this base class because it is the same everywhere :-)